hilti/runtime/include/util.h

```diff
@@ -649,4 +650,97 @@ constexpr std::size_t hashCombine(std::size_t hash1, std::size_t hash2, Hashes..
         return result;
 }
 
+template<typename Data, typename Error>
+class ControlBlockRef;
+
+// Helper class for loosely tracking liveliness of some memory. Code using this
+// would instantiate an instance of this pointing to some data. It can then
+// pass out references to that control block by calling `asRef`. One can get
+// values from these references as long as the original control block is live.
+template<typename Data, typename Error>
+class ControlBlock {
+public:
+    using Ref = ControlBlockRef<Data, Error>;
+
+    ControlBlock() = default;
+    ControlBlock(Data* data) : _data(data) {}
+
+    ControlBlock(const ControlBlock&) = default;
+    ControlBlock(ControlBlock&&) = default;
+
+    ControlBlock& operator=(const ControlBlock&) = default;
+    ControlBlock& operator=(ControlBlock&&) = default;
```

Comment:
I'm wondering if we should `delete` these. Everything still compiles if I do, and conceptually it seems we don't really *want* to create multiple copies of the same block, do we?

Reply:
Hmm, I think we need the copy and move ctrs and assignment operators so we can use these in classes for which we want to support these operations. The implementations for copies should not copy the control block though, otherwise we'd get spooky action at a distance where a iterators into a container copy would get invalidated if the original object gets updated; I'll update these to use a separate `_control`.